webapp/src/cloud.ts

```diff
@@ -0,0 +1,196 @@
+        const result = await auth.apiAsync<CloudProject>(`/api/user/project/${h.id}`);
+        if (result.success) {
+            const userId = auth.user()?.id;
+            const project = result.resp;
+            const header = JSON.parse(project.header);
+            const text = JSON.parse(project.text);
+            const version = project.version;
+            const file: File = {
+                header,
+                text,
+                version
+            };
+            file.header.cloudCurrent = true;
+            file.header.cloudVersion = file.version;
+            file.header.cloudUserId = userId;
+            resolve(file);
+        } else {
+            reject(new Error(result.errmsg));
+        }
+    });
+}
+
+function setAsync(h: Header, prevVersion: Version, text?: ScriptText): Promise<Version> {
+    return new Promise(async (resolve, reject) => {
+        const userId = auth.user()?.id;
+        h.cloudUserId = userId;
+        h.cloudCurrent = false;
+        h.cloudVersion = prevVersion;
+        const project: CloudProject = {
+            id: h.id,
+            header: JSON.stringify(h),
+            text: text ? JSON.stringify(text) : undefined,
+            version: prevVersion
+        }
+        const result = await auth.apiAsync<string>('/api/user/project', project);
+        if (result.success) {
+            h.cloudCurrent = true;
+            h.cloudVersion = result.resp;
+            resolve(result.resp);
+        } else {
+            // TODO: Handle reject due to version conflict
+            reject(new Error(result.errmsg));
+        }
+    });
+}
+
+function deleteAsync(h: Header, prevVersion: Version, text?: ScriptText): Promise<void> {
+    return Promise.resolve();
+}
+
+function resetAsync(): Promise<void> {
+    return Promise.resolve();
+}
+
+export async function syncAsync(): Promise<any> {
+    if (!auth.hasIdentity()) { return; }
+    if (!await auth.loggedIn()) { return; }
+    try {
+        const userId = auth.user()?.id;
+        // Filter to cloud-synced headers owned by the current user.
+        const localCloudHeaders = workspace.getHeaders(true)
+            .filter(h => h.cloudUserId && h.cloudUserId === userId);
+        const remoteHeaders = await listAsync();
+        const remoteHeaderMap = U.toDictionary(remoteHeaders, h => h.id);
+        const tasks = localCloudHeaders.map(async (local) => {
+            const remote = remoteHeaderMap[local.id];
+            delete remoteHeaderMap[local.id];
+            if (remote) {
+                if (local.cloudVersion !== remote.cloudVersion) {
+                    if (local.cloudCurrent) {
+                        // No local changes, download latest.
+                        const file = await getAsync(local);
+                        workspace.saveAsync(file.header, file.text, file.version);
+                    } else {
+                        // Conflict.
+                        // TODO: Figure out how to register these.
```

Comment:
Need to show cloud status on the cards

Reply:
See Cloud Sync OneNote

Reply:
Do we need timestamps or incremental version numbers? Or the user in the loop?
LastWrite wins?

Reply:
hunch: auto-handling merges would be heavier lift and harder to get right than presenting options to user.